File: multi-lang/aoc2022/2/part1/src/main.zig
const std = @import("std");

const input = @embedFile("input");

pub fn main() anyerror!void {
    var lines = std.mem.split(u8, input, "\n");

    var score: u32 = 0;

    while (lines.next()) |line| {
        // newline at end of file
        if (line.len == 0) {
            continue;
        }

        var enemy: u8 = switch (line[0]) {
            // I couldn't figure out how to make it accept "A" => 1
            65 => 1,
            66 => 2,
            67 => 3,
            else => unreachable,
        };

        var self: u8 = switch (line[2]) {
            88 => 1,
            89 => 2,
            90 => 3,
            else => unreachable,
        };

        // won
        if ((enemy == 3 and self == 1) or (enemy < self and !(enemy == 1 and self == 3))) {
            score += 6;
        }
        // tie
        else if (enemy == self) {
            score += 3;
        }
        score += self;
    }

    std.log.info("score = {d}", .{score});
}

// a: rock
// b: paper
// c: scissors
//
// x: rock
// y: paper
// z: scissors
//
// rock paper -> win 6
// paper rock -> loss 0
// scissors scissors -> draw 3
//
// 1 2 
// 2 1
// 3 3
//
// enemy self
// self > enemy -> win
// self < enemy -> loss
// self == enemy -> tie
//
// special 3 1 
// 1 3 -> loss
// 3 1 -> win

File: multi-lang/aoc2022/2/part1/rust/src/main.rs
fn main() {
    let input = include_str!("input");

    let mut score: u32 = 0;

    for line in input.split("\n") {
        if line.len() == 0 {
            continue;
        }

        let enemy = match line.chars().nth(0).unwrap() {
            'A' => 1,
            'B' => 2,
            'C' => 3,
            _ => panic!("Unreachable")
        };

        let me = match line.chars().nth(2).unwrap() {
            'X' => 1,
            'Y' => 2,
            'Z' => 3,
            _ => panic!("Unreachable")
        };

        // won
        if (enemy == 3 && me == 1) || (enemy < me && !(enemy == 1 && me == 3)) {
            score += 6;
        }
        // tie
        else if enemy == me {
            score += 3;
        }
        score += me;
    }

    println!("score = {}", score);
}
